rfc/src/rfcs/0010-rust-ub-checks.md

```diff
@@ -0,0 +1,175 @@
+non-determinism, such as CBMC's deallocated check.
+See [0010-rust-ub-checks.md#open-questions] for alternatives to be considered.
+
+### Concrete playback of UB checks
+
+A counter example that triggers UB may require extra tooling for debugging, e.g.: MIRI or valgrind.
+We propose that tests added by concrete playback in those cases should include a comment stating such limitation.
+Other possibilities are discussed in [0010-rust-ub-checks.md#open-questions].
+
+### Delayed undefined behavior
+
+In some cases, undefined behavior may occur in safe code that interacts with unsafe code.
+
+For example, the code below shows how reading a boolean value in safe code may trigger UB due to an
+invalid write:
+
+```rust
+#[kani::proof]
+fn buggy_code() {
+    let mut value: bool = false;
+    unsafe {
+        let ptr: *mut u8 = &mut value as *mut _ as *mut u8;
+        *ptr = 8; // This assignment does not cause UB...
+    }
+    assert!(value); // ...but this read triggers UB! ⚠️
+}
+```
+
+It may not be scalable for Kani to verify that every value produced is valid.
+In cases like that, we may opt for implementing one of the following UB checks:
+
+1. Perform static analysis to first identify which code may be affected by unsafe code, and only instrument those that
+   are reachable from the unsafe code.
+2. Add eager UB detection in unsafe code that will fail verification if there are side effects that we may not be
+   able to detect efficiently. For example, we could fail if casting between types that have different value
+   requirements, i.e.: fail in the cast `*mut bool as *mut u8`.
+
+For option 2, Kani may fail verification even if the code does not have UB.
+In those cases, the Kani check must be explicit about this failure being an over-approximation.
+For that, we propose that the check message is clear about this limitation, and potentially its status
+(see [0010-rust-ub-checks.md#open-questions]).
+
+### Opt-out a class of UB checks
+
+Undefined behavior checks that have been stabilized should be added by default by Kani.
+However, Kani should provide a fine-grained control for users to choose which checks to disable.
+
+For every class of UB checks to be added, we shall add a command line argument to disable the check for all harnesses,
+as well as an attribute that would allow users to opt-out a check for specific harnesses.
+
+Command line arguments should follow the existing name schema:
+`--no-[CHECK-NAME]-checks` where the CHECK-NAME is kebab case. There is no need to add `--[CHECK-NAME]-checks` since
+they are enabled by default.
+
+We propose to add the following attribute:
+
+```rust
+#[kani::proof]
+#[kani::disable(CHECK_NAMES)]
+fn harness() {}
+```
+
+Where `CHECK_NAMES` is a list of check category names in snake case, such as `memory_safety`, `valid_values`.
+
+For checks that are unstable, Kani should support both opt-out mechanisms.
```

Comment:
Why this requirement? I think it'll be trivial to support opt-outs once the groundwork is in place, but otherwise it's putting obstacles to unstable features.

Reply:
Removed this